src/engine/thread.js

```diff
@@ -27,92 +33,70 @@ class _StackFrame {
         this.warpMode = warpMode;
 
         /**
-         * Reported value from just executed block.
-         * @type {Any}
-         */
-        this.justReported = null;
-
-        /**
-         * The active block that is waiting on a promise.
-         * @type {string}
-         */
-        this.reporting = '';
-
-        /**
-         * Persists reported inputs during async block.
+         * Procedure parameters.
          * @type {Object}
          */
-        this.reported = null;
+        this.params = params;
 
         /**
-         * Name of waiting reporter.
-         * @type {string}
-         */
-        this.waitingReporter = null;
-
-        /**
-         * Procedure parameters.
+         * A context passed to block implementations.
          * @type {Object}
          */
-        this.params = null;
+        this.executionContext = {};
 
         /**
-         * A context passed to block implementations.
-         * @type {Object}
+         * Has this frame changed and need a reset?
+         * @type {boolean}
          */
-        this.executionContext = null;
+        this.needReset = false;
     }
 
     /**
-     * Reset all properties of the frame to pristine null and false states.
-     * Used to recycle.
+     * Reset some properties of the frame to default values. Used to recycle.
      * @return {_StackFrame} this
      */
     reset () {
-
         this.isLoop = false;
-        this.warpMode = false;
-        this.justReported = null;
-        this.reported = null;
-        this.waitingReporter = null;
-        this.params = null;
-        this.executionContext = null;
+        this.executionContext = {};
+        this.needReset = false;
 
         return this;
     }
 
     /**
      * Reuse an active stack frame in the stack.
-     * @param {?boolean} warpMode defaults to current warpMode
-     * @returns {_StackFrame} this
      */
-    reuse (warpMode = this.warpMode) {
-        this.reset();
-        this.warpMode = Boolean(warpMode);
-        return this;
+    reuse () {
```

Comment:
why not just call reset here?

Reply:
Performance. I think reset and reuse need to be fast. So I inlined the new reset here. I think it would be better for us to get rid of one of reset or reuse then call one from the other.